prbt_hardware_support/include/prbt_hardware_support/adapter_sto.h

```diff
@@ -100,9 +143,201 @@ class AdapterSto
    * before a driver disable enables the breaks.
    */
   static constexpr int DURATION_BETWEEN_HOLD_AND_DISABLE_MS {200};
-
-  static constexpr double WAIT_FOR_SERVICE_TIMEOUT_S {5.0};
 };
 
+//! Simple typedef for class like usage
+typedef AdapterStoTemplated<> AdapterSto;
+
+template <class T>
+const std::string AdapterStoTemplated<T>::HOLD_SERVICE{"manipulator_joint_trajectory_controller/hold"};
+
+template <class T>
+const std::string AdapterStoTemplated<T>::UNHOLD_SERVICE{"manipulator_joint_trajectory_controller/unhold"};
+
+template <class T>
+const std::string AdapterStoTemplated<T>::RECOVER_SERVICE{"driver/recover"};
+
+template <class T>
+const std::string AdapterStoTemplated<T>::HALT_SERVICE{"driver/halt"};
+
+template <class T>
+AdapterStoTemplated<T>::AdapterStoTemplated(std::function<T(std::string)> create_service_client)
+    : hold_srv_client_(create_service_client(HOLD_SERVICE)),
+      unhold_srv_client_(create_service_client(UNHOLD_SERVICE)),
+      recover_srv_client_(create_service_client(RECOVER_SERVICE)),
+      halt_srv_client_(create_service_client(HALT_SERVICE))
+{
+}
+
+template <class T>
+void AdapterStoTemplated<T>::run()
+{
+  Action current_action{Action::HALTING};
+  bool execute_action{false};
+
+  ros::Rate rate(10);
+  while (!terminate_)
+  {
+    {
+      std::lock_guard<std::mutex> lock(action_mutex_);
+      execute_action = execute_action || (next_action_ != current_action);
+      if (execute_action)
+      {
+        current_action = next_action_;
+      }
+    }
+
+    if (execute_action)
+    {
+      bool exec_result = execute(current_action);
+      execute_action = !exec_result;
+      autoUpdate(current_action, exec_result);
+    }
+
+    rate.sleep();
+  }
+}
+
+template <class T>
+AdapterStoTemplated<T>::~AdapterStoTemplated()
+{
+  terminate();
+}
+
+template <class T>
+void AdapterStoTemplated<T>::runAsync()
+{
+  run_thread_ = std::thread{[this] { this->run(); }};
+}
+
+template <class T>
+void AdapterStoTemplated<T>::updateSto(const bool STO)
+{
+  std::lock_guard<std::mutex> lock(action_mutex_);
+
+  if (STO)
+  {
+    if (next_action_ == Action::HALTING)
+    {
+      next_action_ = Action::RECOVERING;
```

Comment:
The transition HALTING->RECOVERING isn't supported.
Please also correct diag_state_sto_executor_action.png

Reply:
What do you mean? This triggers recover when STO false -> true.

Reply:
Exactly. But the shortcut is only desired for RECOVERING->HALTING. Once STO goes to false, the full halt is always executed (see doors specification of PSS4000 function block).

Reply:
Sorry, if the code is misleading, I am rewriting this part at the moment and hope to resolve the ambiguities.

The full halt (when STO goes false) is UNHOLDING -> HOLDING -> HALTING

HALTING -> RECOVERING for STO true is no shortcut

Reply:
Or: Do you deliberately deviate from the PSS4000 spec and leave that "full stop" to the safety controller? As said: In my opinion the diag_state_sto_executor_action.png should be asymmetric: Only having one shortcut for RECOVERING->HALTING. In state HOLDING, the sto=true could be visualized as self-transition / effectively a no-op.

Reply:
Ok, I got it. Thanks!